File: src/game/__tests__/combat.test.ts
import { describe, it, expect } from "vitest";
import { buildEnemyUnits, simulateBattle, unitFromInstance } from "../combat";
import type { HeroInstance } from "../../types";
import { HEROES } from "../heroes";

function makeHero(
  templateId: string,
  overrides: Partial<HeroInstance> = {}
): HeroInstance {
  return {
    instanceId: `t-${templateId}`,
    templateId,
    level: 20,
    stars: 5,
    xp: 0,
    echoStacks: 0,
    gearTier: 2,
    ...overrides,
  };
}

describe("combat", () => {
  it("produces a winner within 40 turns", () => {
    const allies = HEROES.slice(0, 5).map((h, i) =>
      unitFromInstance(makeHero(h.id), "ally", i)
    );
    const enemies = buildEnemyUnits(1, 1);
    const r = simulateBattle(allies, enemies);
    expect(["ally", "enemy"]).toContain(r.winner);
    expect(r.turns).toBeLessThanOrEqual(40);
  });

  it("scaling enemies win against level-1 heroes eventually", () => {
    const weakHeroes = HEROES.slice(0, 5).map((h, i) =>
      unitFromInstance(makeHero(h.id, { level: 1, stars: 3, gearTier: 0 }), "ally", i)
    );
    const toughEnemies = buildEnemyUnits(8, 8);
    const r = simulateBattle(weakHeroes, toughEnemies);
    expect(r.winner).toBe("enemy");
  });
});

File: android/app/src/test/java/com/herobrawl/game/engine/CombatTest.kt
package com.herobrawl.game.engine

import com.herobrawl.game.data.Heroes
import com.herobrawl.game.model.HeroInstance
import org.junit.Assert.assertTrue
import org.junit.Test
import kotlin.random.Random

class CombatTest {
    private fun hero(id: String, level: Int = 20, stars: Int = 5, gear: Int = 2) =
        HeroInstance(
            instanceId = "t-$id", templateId = id, level = level,
            stars = stars, xp = 0, echoStacks = 0, gearTier = gear,
        )

    @Test
    fun `battle produces a winner within 40 turns`() {
        val allies = Heroes.all.take(5).mapIndexed { i, h ->
            Combat.unitFromInstance(hero(h.id), com.herobrawl.game.engine.BattleUnit.Side.ALLY, i, 0.0, 0.0, 0.0)
        }
        val enemies = Combat.buildEnemyUnits(1, 1, rnd = Random(1))
        val r = Combat.simulate(allies, enemies, rnd = Random(2))
        assertTrue(r.turns <= 40)
    }

    @Test
    fun `weak heroes lose to late stage enemies`() {
        val allies = Heroes.all.take(5).mapIndexed { i, h ->
            Combat.unitFromInstance(
                hero(h.id, level = 1, stars = 3, gear = 0),
                com.herobrawl.game.engine.BattleUnit.Side.ALLY, i, 0.0, 0.0, 0.0
            )
        }
        val enemies = Combat.buildEnemyUnits(8, 8, rnd = Random(1))
        val r = Combat.simulate(allies, enemies, rnd = Random(2))
        assertTrue(r.winner == com.herobrawl.game.engine.BattleUnit.Side.ENEMY)
    }
}
